### PR TITLE
Inject job streamer in the gateway

### LINKED ISSUE #12388: Provide ClientStreamer instance to the GatewayGrpcService/EndpointManager
**Description**

Before the ZPA team gets started on their side in the gateway, they should already have access to a provisioned `ClientStreamer` which they can use to add/remove streams in order to forward activated jobs.

It's perfectly fine if the streamer is initially unused, as we'll likely get to this before them, and the goal here is simply to make sure they're unblocked.

### PR DESCRIPTION
**Pull request by npepinpe**
## Description

This PR adds job specific implementations/adapters around the remote streaming client capabilities of the transport module. As an end result, there is a `ClientStreamer<JobActivationProperties>` available to the `EndpointManager`, ready to be used by the ZPA team when implementing the job streaming endpoint.

There aren't really any tests, as there isn't a way to test the integration of all these things yet without the gRPC API. Once that's present, we can add integration tests.

One other caveat is I had to introduce a new `ClientStreamService` interface in the transport to avoid exposing that the streamer implementation is an actor itself. Right now there's only one implementation, but I would like to do a follow up pass after this to clean things up a little. I'm also not super happy with the startup process of having to wait for the broker client's topology manager to start before adding the listener, but that can go in a follow up as well.

Another issue is around the integration tests where we create the gateway. Adding more things to construct a `Gateway` instance meant updating multiple tests. I'd like to also time box some time as a follow up to figure out how we can simplify this, e.g. a gateway analogous rule like `EmbeddedBrokerRule` or something. There's a bunch of duplication around constructing a `Gateway` instance.

@koevskinikola As a caveat, I ended up duplicating the `JobActivationProperties` type in the gateway module as well, since the gateway is not aware of the engine. I think that interface and its implementation should be moved to the `protocol-impl`, but I'll let your team do that to avoid conflicts. When that's done, just remove the duplicate implementation.

## Related issues

closes #12388 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Other teams:
If the change impacts another team an issue has been created for this team, explaining what they need to do to support this change.
- [ ] [Operate](https://github.com/camunda/operate/issues)
- [ ] [Tasklist](https://github.com/camunda/tasklist/issues)
- [ ] [Web Modeler](https://github.com/camunda/web-modeler/issues)
- [ ] [Desktop Modeler](https://github.com/camunda/camunda-modeler/issues)
- [ ] [Optimize](https://github.com/camunda/camunda-optimize/issues)

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

**Comment by npepinpe** [comment]
Follow ups:

- [ ] Refactor gateway creation in tests to a utility class
- [ ] Revisit `ClientServiceImpl`
- [ ] Revisit how the job stream client is created and its inter-dependency with the topology manager
- [ ] Remove circular dependency between `ClientStream` implementations and the `AggregatedStream`

**Comment by npepinpe** [comment]
@deepthidevaki - as noted above, I've identified some follow ups, but we could also do them here. I would like them as follow ups because

1. Avoid making the PR too big (e.g. adding a new gateway rule like the `EmbeddedBrokerRule` would add a lot, and while useful, isn't strictly required here)
2. Taking a step back and revisiting this with fresh eyes/perspective (plus maybe when it's actually in use) may help when it comes to figuring out the right design (e.g. service and streamer sharing the implementation, or should the service be an actor or just receive a `ConcurrencyControl` instead, etc.).

File: broker/src/main/java/io/camunda/zeebe/broker/system/EmbeddedGatewayService.java [hunk 1]
```diff
@@ -7,48 +7,71 @@
...
   }
 
   public ActorFuture<Gateway> start() {
-    brokerClient.start();
+    jobStreamClient.start();
+
+    // before we can add the job stream client as a topology listener, we need to wait for the
+    // topology to be set up, otherwise the callback may be lost
+    final var brokerClientStart =
+        brokerClient.start().stream().collect(new ActorFutureCollector<>(concurrencyControl));
```
Comment (deepthidevaki) [inline broker/src/main/java/io/camunda/zeebe/broker/system/EmbeddedGatewayService.java:70]: :question: Should we also add the future returned by `jobStreamClient.start()` to wait until jobStreamClient is started before adding it as topology listener?

File: dist/src/main/java/io/camunda/zeebe/gateway/StandaloneGateway.java [hunk 2]
```diff
@@ -94,8 +99,16 @@ public void run(final String... args) throws Exception {
...
 
-    gateway = new Gateway(configuration, brokerClient, actorScheduler);
+    actorScheduler.start();
+    atomixCluster.start();
+    jobStreamClient.start();
+
+    // before we can add the job stream client as a topology listener, we need to wait for the
+    // topology to be set up, otherwise the callback may be lost
+    brokerClient.start().forEach(ActorFuture::join);
+    brokerClient.getTopologyManager().addTopologyListener(jobStreamClient);
```
Comment (deepthidevaki) [inline dist/src/main/java/io/camunda/zeebe/gateway/StandaloneGateway.java:109]: :question: Same as above. We should also wait until jobStreamClient is started.

**Review by deepthidevaki — APPROVED** [review]
:rocket: Please check my comments before merging.

File: broker/src/main/java/io/camunda/zeebe/broker/system/EmbeddedGatewayService.java (hunk 1 above)
Comment (npepinpe) [reply broker/src/main/java/io/camunda/zeebe/broker/system/EmbeddedGatewayService.java:70]: Ah, yes, I added it initially when the start of the job stream client was synchronous :see_no_evil:

**Comment by npepinpe** [comment]
bors merge

**Pull request merged**